chatsky/context_storages/database.py

```diff
@@ -8,181 +8,250 @@
+        """`full_path` without a prefix defining db used."""
+        self.rewrite_existing = rewrite_existing
+        """Whether to rewrite existing data in the storage."""
+        self._subscripts = dict()
+        self._sync_lock = Lock()
+        self.connected = False
+        for field in (NameConfig._labels_field, NameConfig._requests_field, NameConfig._responses_field):
+            value = configuration.get(field, self._default_subscript_value)
+            if (not isinstance(value, int)) or value >= 1:
+                self._subscripts[field] = value
+            else:
+                raise ValueError(f"Invalid subscript value ({value}) for field {field}")
+
+    @property
+    @abstractmethod
+    def is_concurrent(self) -> bool:
+        raise NotImplementedError
 
-    def __getitem__(self, key: Hashable) -> Context:
-        """
-        Synchronous method for accessing stored Context.
+    @staticmethod
+    def _lock(function: Callable[..., Awaitable[Any]]):
+        @wraps(function)
+        async def wrapped(self, *args, **kwargs):
+            if not self.is_concurrent:
+                async with self._sync_lock:
+                    return await function(self, *args, **kwargs)
+            else:
+                return await function(self, *args, **kwargs)
 
-        :param key: Hashable key used to store Context instance.
-        :return: The stored context, associated with the given key.
-        """
-        return asyncio.run(self.get_item_async(key))
+        return wrapped
 
-    @abstractmethod
-    async def get_item_async(self, key: Hashable) -> Context:
-        """
-        Asynchronous method for accessing stored Context.
+    @classmethod
+    def _validate_field_name(cls, field_name: str) -> str:
+        if field_name not in (NameConfig._labels_field, NameConfig._requests_field, NameConfig._responses_field):
+            raise ValueError(f"Invalid value '{field_name}' for argument 'field_name'!")
+        else:
+            return field_name
 
-        :param key: Hashable key used to store Context instance.
-        :return: The stored context, associated with the given key.
-        """
+    async def connect(self) -> None:
+        self.connected = True
+
+    @abstractmethod
+    async def _load_main_info(self, ctx_id: str) -> Optional[Tuple[int, int, int, bytes, bytes]]:
         raise NotImplementedError
 
-    def __setitem__(self, key: Hashable, value: Context):
+    @_lock
+    async def load_main_info(self, ctx_id: str) -> Optional[Tuple[int, int, int, bytes, bytes]]:
         """
-        Synchronous method for storing Context.
-
-        :param key: Hashable key used to store Context instance.
-        :param value: Context to store.
+        Load main information about the context.
         """
-        return asyncio.run(self.set_item_async(key, value))
+        if not self.connected:
+            logger.debug(f"Connecting to context storage {type(self).__name__} ...")
```

Comment:
I think we should do an info level log in the `connect` method about the connection being made and here do a warning level log that db was not initiated via `pipeline.run`.

See my reply to [the connection thread](https://github.com/deeppavlov/chatsky/pull/93/files#r1851023243) for more context.